[Shipping labels] Add missing tracks

irfano opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the Linear issue this PR addresses, e.g., WOOMOB-373. -->
Closes WOOMOB-684

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR adds missing events for shipping labels and improves error tracking. The changes in this PR are based on feedback from #14306.

### Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->
#### wcs_package_selection_step
1. Open create shipping label screen.
2. Tap "Select a package"
3. Fill "Length", "Width" and "Height" fields.
4. Check "Save this as a new package template".
5. Fill "Weight", and "Package Name" fields.
6. Tap Add Package button.
7. `🔵 Tracked: woocommerceandroid_wcs_package_selection_step, Properties {"state":"saving_success"`
8. For error, follow steps in https://github.com/woocommerce/woocommerce-android/pull/14295.
9. Repeat steps 1-7.
10. `🔵 Tracked: woocommerceandroid_wcs_package_selection_step, Properties: {"state":"saving_failed","error":"API_ERROR"`
11. Go to Carrier tab.
12. Unstar a package.
13. `🔵 Tracked: woocommerceandroid_wcs_package_selection_step, Properties: {"state":"removing_failed"`
14. Go to Saved tab.
15. Remove a package.
16. `🔵 Tracked: woocommerceandroid_wcs_package_selection_step, Properties: {"state":"removing_failed"`

#### wcs_rate_selection_step
1. Open create shipping label screen.
2. Select a package.
3. Type 9999999999 weight.
4. `🔵 Tracked: woocommerceandroid_wcs_rate_selection_step, Properties: {"state":"loading_failed","error":"no_rates_available"`

### The tests that have been performed
<!-- To give the reviewer an idea of what could be missed in terms of testing -->
Steps above

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

irfano committed:
Fix error tracking for package saving and removal failures

irfano committed:
Add saving event for custom shipping label packages

irfano committed:
Add no_rates_available error message to WCS_RATE_SELECTION_STEP event

irfano committed:
Use mocked tracker in WooShippingLabelPackageCreationViewModelTest

irfano committed:
Add unit test to verify tracking event with removing_failed property

irfano committed:
Add unit tests for events triggered after saving custom packages

irfano committed:
Improve error tracking for WCS_PACKAGE_SELECTION_STEP event

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-android/milestone/262">22.8 ❄️</a>. The due date for this milestone has already passed.<br />
Please assign it to a milestone with a later deadline or check whether the release for this milestone has already been finished.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

irfano committed:
Fix unit test for failure when unstarring a package

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/WooShippingLabelCreationViewModel.kt:
```diff
@@ -596,7 +596,9 @@ class WooShippingLabelCreationViewModel @Inject constructor(
-                    trackShippingRatesLoading(isSuccess = false, error = shippingRatesResult.exceptionOrNull()?.message)
+                    val error = shippingRatesResult.exceptionOrNull()?.message
+                        ?: if (shippingRatesResult.getOrNull()?.isEmpty() == true) "no_rates_available" else null
+                    trackShippingRatesLoading(isSuccess = false, error = error)
```
np, just sharing a suggestion for a separate PR, not this one, I think we can improve this and also the `GetShippingRates` using the built-in helpers of the `Result` class:
- We can simplify `getShippingRates` logic using `Result#map`.
- We can simplify the logic here using `Result#fold`

I would also update `GetShippingRates` to return an error when the rates are empty, instead of having the logic here in the ViewModel, the `ViewModel` is already complex, and moving more logic to the Use cases would simplify it further.

irfano replied:
These were good suggestions. I've updated `GetShippingRates` and `WooShippingLabelCreationViewModel`: https://github.com/woocommerce/woocommerce-android/pull/14323/commits/4216b383c1ec1197d4345e8694690b88fdee022d

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/packages/WooShippingLabelPackageCreationViewModel.kt:
```diff
@@ -131,7 +132,10 @@ class WooShippingLabelPackageCreationViewModel @Inject constructor(
-                        mapOf(KEY_STATE to "removing_failed", KEY_ERROR to it.exceptionOrNull()?.message.orEmpty())
+                        mapOf(
+                            KEY_STATE to "removing_failed",
+                            KEY_ERROR to (it.exceptionOrNull() as? WooException)?.error?.type?.name.orEmpty()
```
Just trying to understand here, why sometimes we track the error message as the `error` and sometimes the error type like here?
Not necessarily in this PR, but IMO it would be better to update the error reporting to use the `error_context`, `error_description` and `error_type` as discussed in Slack, as then the `error_type` can contain the `error.type.name` and the `error_description` can contain the `error.message`.

hichamboushaba replied:
Just adding one point here, from my personal experience, sometimes we need both the `type` and `message` to understand the cause when investigating some issues.

irfano replied:
Uses cases don't have consistent error types, sometimes they throw `Exception`, sometimes `WooException`. I didn’t include refactoring use cases in the analytics PRs. I checked whether message or type had more meaningful content in each case, which is why I used one or the other accordingly.

> Not necessarily in this PR, but IMO it would be better to update the error reporting to use the error_context, error_description and error_type as discussed in Slack, as then the error_type can contain the error.type.name and the error_description can contain the error.message.

I agree that we should use `error_context`, `error_description` and `error_type` for consistency. Shipping labels isn’t the only feature that uses just the error property, so I think this is a low-priority issue.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/packages/WooShippingLabelPackageCreationViewModel.kt:
```diff
@@ -297,7 +306,7 @@ class WooShippingLabelPackageCreationViewModel @Inject constructor(
             ?.let { PackageData.fromPackageDAO(it) }
             ?.let { Result.success(it) }
-            ?: Result.failure(Throwable("Failed to save package"))
+            ?: Result.failure(Throwable(response.error.type.toString()))
```
I think using `WooException` here would be better as it exposes both the error type and message.

```suggestion
            ?: Result.failure(WooException(response.error))
```

irfano replied:
For now, I only need the error type for tracking. The original parameter for `Result.failure` is `Throwable`. If I use `WooException` here, I would need to cast it to `WooException` to access the type. Additionally, the consumer function would need to be aware that `sendCustomPackageToStore()` returns a `WooException`. I think using `Throwable` is a simpler approach in this case.

hichamboushaba replied:
> Additionally, the consumer function would need to be aware that sendCustomPackageToStore() returns a WooException

Why? We use `fold` in the consumer function, and and this will continue to work, `WooException` is a `Throwable`.

> For now, I only need the error type for tracking

This is related to my other [question](https://github.com/woocommerce/woocommerce-android/pull/14323#discussion_r2207740096), it seems to me we report the [`message`](https://github.com/woocommerce/woocommerce-android/blob/4216b383c1ec1197d4345e8694690b88fdee022d/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/WooShippingLabelCreationViewModel.kt#L600) in some places, and the `type` in other places.

irfano replied:
> Why? We use fold in the consumer function, and and this will continue to work, WooException is a Throwable.

Yes, but `Throwable` doesn't have a `type` property. To access the `type` from `WooException`, I'd need to cast `Throwable` to `WooException`. That said, your suggestion will eventually be necessary when we start including more error data in events. So, I'll leave using `WooException` for PR where we add `error_context`, `error_description` and `error_type`.

hichamboushaba reviewed (approved):
Good job @irfano, I left some comments, but I don't think they are blockers for now, we can maybe implement them in a separate PR if you prefer so.

irfano committed:
Refactor shipping rates fetching and error handling

This commit refactors the `GetShippingRates` use case and `WooShippingLabelCreationViewModel` to improve how shipping rates are fetched and how errors are handled.

irfano committed:
Add test for handling empty shipping rates response
